storage/queue.go

```diff
@@ -55,16 +55,44 @@ type FlushFunc func(ctx context.Context, entries []tessera.Entry) (index uint64,
 // The provided FlushFunc will be called with a slice containing the contents of the queue, in
 // the same order as they were added, when either the oldest entry in the queue has been there
 // for maxAge, or the size of the queue reaches maxSize.
-func NewQueue(maxAge time.Duration, maxSize uint, f FlushFunc) *Queue {
+func NewQueue(ctx context.Context, maxAge time.Duration, maxSize uint, f FlushFunc) *Queue {
 	q := &Queue{
 		flush:    f,
 		inFlight: make(map[string]*entry, maxSize),
 	}
+
+	// The underlying queue implementation blocks additions during a flush.
+	// This blocks the filling of the next batch unnecessarily, so we'll
+	// decouple the queue flush and storage write by handling the later in
+	// a worker goroutine.
+	// This same worker thread will also handle the callbacks to f.
+	work := make(chan []*entry, 1)
+	toWork := func(items []interface{}) {
```

Comment:
Why is the `interface{}` instead of `*entry` used here?

Reply:
That's an artifact of the underlying queue library, it's pre-generics (I guess) so is built as a queue of `interface{}`. This func just converts those back to the actual type we know it contains when the queue flushes.